rotkehlchen/chain/ethereum/modules/weth/decoder.py

```diff
@@ -0,0 +1,153 @@
+            return self._decode_withdrawal_event(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+
+        return None, None
+
+    def _decode_deposit_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        depositor = hex_or_bytes_to_address(tx_log.topics[1])
+        deposited_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        deposited_amount = asset_normalized_value(amount=deposited_amount_raw, asset=self.eth)
+
+        for event in decoded_events:
+            if event.event_type == HistoryEventType.SPEND and event.location_label == depositor and event.balance.amount == deposited_amount and event.asset == self.eth:  # noqa: E501
+                event.counterparty = CPT_WETH
+                event.event_type = HistoryEventType.DEPOSIT
+                event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                event.notes = f'Deposit {deposited_amount} {self.eth.symbol} for WETH'  # noqa: E501
+
+        event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=self.base_tools.get_next_sequence_counter(),
+            timestamp=ts_sec_to_ms(transaction.timestamp),
+            location=Location.BLOCKCHAIN,
+            event_type=HistoryEventType.RECEIVE,
+            event_subtype=HistoryEventSubType.RECEIVE_WRAPPED,
+            asset=A_WETH,
+            balance=Balance(amount=deposited_amount),
+            location_label=depositor,
+            counterparty=CPT_WETH,
+            notes=f'Receive {deposited_amount} {self.weth.symbol}',
+        )
+        return event, None
+
+    def _decode_withdrawal_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        withdrawer = hex_or_bytes_to_address(tx_log.topics[1])
+        withdrawed_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        withdrawed_amount = asset_normalized_value(amount=withdrawed_amount_raw, asset=self.eth)
+
+        in_event = None
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.counterparty in (WETH_CONTRACT, withdrawer) and
```

Comment:
why is this in needed? Shouldn't be always the weth contract?

Reply:
Not in all cases. If it's called from a protocol, the counterparty is the protocol

Reply:
Yes, I simply ignore.